### PR TITLE
Handle new `source(…)`

### PR DESCRIPTION
**Pull request by RobinMalfait**
This PR introduces a new `source(…)` argument and improves on the existing `@source`. The goal of this PR is to make the automatic source detection configurable, let's dig in.

By default, we will perform automatic source detection starting at the current working directory. Auto source detection will find plain text files (no binaries, images, ...) and will ignore git-ignored files.

If you want to start from a different directory, you can use the new `source(…)` next to the `@import "tailwindcss/utilities" layer(utilities) source(…)`.

E.g.:

```css
/* ./src/styles/index.css */
@import 'tailwindcss/utilities' layer(utilities) source('../../');
```

Most people won't split their source files, and will just use the simple `@import "tailwindcss";`, because of this reason, you can use `source(…)` on the import as well:

E.g.:

```css
/* ./src/styles/index.css */
@import 'tailwindcss' source('../../');
```

Sometimes, you want to rely on auto source detection, but also want to look in another directory for source files. In this case, yuo can use the `@source` directive:

```css
/* ./src/index.css */
@import 'tailwindcss';

/* Look for `blade.php` files in `../resources/views` */
@source '../resources/views/**/*.blade.php';
```

However, you don't need to specify the extension, instead you can just point the directory and all the same automatic source detection rules will apply.

```css
/* ./src/index.css */
@import 'tailwindcss';

@source '../resources/views';
```

If, for whatever reason, you want to disable the default source detection feature entirely, and only want to rely on very specific glob patterns you define, then you can disable it via `source(none)`.

```css
/* Completely disable the default auto source detection */
@import 'tailwindcss' source(none);

/* Only look at .blade.php files, nothing else  */
@source "../resources/views/**/*.blade.php";
```

Note: even with `source(none)`, if your `@source` points to a directory, then auto source detection will still be performed in that directory. If you don't want that, then you can simply add explicit files in the globs as seen in the previous example.

```css
/* Completely disable the default auto source detection */
@import 'tailwindcss' source(none);

/* Run auto source detection in `../resources/views` */
@source "../resources/views";
```

### CODE REVIEW

**Comment by RobinMalfait** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14767?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14767** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14767?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#14766** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14766?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#14765** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14765?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @RobinMalfait and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/oxide/src/lib.rs [hunk 1]
```diff
@@ -219,6 +219,36 @@ impl Scanner {
...
+
+        // Find all `@source` globs that point to a directory. If so, promote the source to auto
+        // source detection instead.
+        if let Some(sources) = &mut self.sources {
+            for source in sources {
+                // If a glob ends with `**/*`, then we just want to register the base path as a new
+                // base.
+                if source.pattern.ends_with("**/*") {
+                    source.pattern = source.pattern.trim_end_matches("**/*").to_owned();
+                }
```
Comment (thecrypticace) [inline crates/oxide/src/lib.rs:231]: 1. `trim_end_matches` can be done unconditionally — no need to check `ends_with` first. I'd imagine the compiler is smart enough to omit the extra heap allocation in optimized builds.
2. `trim_end_matches` removes every occurrence which I think is fine. But the presence of the `ends_with` makes me think we wanted to do `strip_suffix` instead.

File: crates/oxide/src/lib.rs [hunk 2]
```diff
@@ -219,6 +219,36 @@ impl Scanner {
...
+
+                let path = PathBuf::from(&source.base).join(&source.pattern);
+                let Some(folder_name) = path.file_name() else {
+                    continue;
+                };
+
+                // Contains a file extension, e.g.: `foo.html`, therefore we don't want to
+                // detect sources here.
+                if folder_name.to_str().unwrap().contains(".") {
```
Comment (thecrypticace) [inline crates/oxide/src/lib.rs:240]: While this is likely to be correct _most_ of the time — folder names can definitely have dots in them. I think the correct check here requires using the filesystem.

File: crates/oxide/src/lib.rs [hunk 3]
```diff
@@ -231,7 +218,58 @@ impl Scanner {
...
+                };
+
+                expression
+                    .into_iter()
+                    .filter_map(|expanded_pattern| expanded_pattern.map(|x| x.into()).ok())
+                    .map(move |pattern| GlobEntry {
+                        base: source.base.clone(),
+                        pattern,
+                    })
+                    .collect::<Vec<_>>()
```
Comment (thecrypticace) [inline crates/oxide/src/lib.rs:237]:
```suggestion
                expression
                    .into_iter()
                    .filter_map(Result::ok)
                    .map(move |pattern| GlobEntry {
                        base: source.base.clone(),
                        pattern: pattern.into(),
                    })
                    .collect::<Vec<_>>()
```

File: packages/@tailwindcss-vite/src/index.ts [hunk 4]
```diff
@@ -379,9 +384,26 @@ class Root {
...
+        // Disable auto source detection
+        if (this.compiler.root === 'none') {
+          this.includeCandidatesFromModuleGraph = false
+          return []
+        }
+
+        // No root specified, use the module graph
+        if (this.compiler.root === null) {
+          return []
+        }
```
Comment (thecrypticace) [inline packages/@tailwindcss-vite/src/index.ts:397]:
```suggestion
        if (this.compiler.root === null) {
          this.includeCandidatesFromModuleGraph = false

          return []
        }
```

File: packages/@tailwindcss-vite/src/index.ts [hunk 5]
```diff
@@ -379,9 +384,26 @@ class Root {
...
+
+        // TODO: In a follow up PR we want this filter this against the module graph.
+        this.includeCandidatesFromModuleGraph = false
```
Comment (thecrypticace) [inline packages/@tailwindcss-vite/src/index.ts:400]:
```suggestion
        this.includeCandidatesFromModuleGraph = true
```

**Comment by RobinMalfait** [comment]
closed in favor of https://github.com/tailwindlabs/tailwindcss/pull/14820

**Pull request closed without merging**